internal/datacoord/snapshot_meta.go

```diff
@@ -0,0 +1,611 @@
+//
+// ARCHITECTURE:
+// - snapshotMeta: Central metadata manager coordinating catalog persistence and S3 storage
+// - SnapshotDataInfo: In-memory cache of snapshot metadata with fast segment/index lookup
+// - Catalog: Persistent storage of snapshot metadata in etcd/KV store
+// - Reader/Writer: S3 operations for reading/writing complete snapshot data
+//
+// METADATA ORGANIZATION:
+// - Memory: snapshotID -> SnapshotDataInfo (with segment/index ID sets for fast lookup)
+// - Catalog (etcd): Collection snapshots metadata (name, ID, collection, partitions, S3 location)
+// - S3: Complete snapshot data (segments, indexes, schema) at S3Location path
+//
+// LIFECYCLE:
+// 1. Create: Save complete snapshot data to S3, then save metadata to catalog
+// 2. Read: Get snapshot info from memory, read complete data from S3 if needed
+// 3. List: Query memory cache with optional collection/partition filtering
+// 4. Drop: Remove from memory, delete from catalog, delete from S3
+// 5. Reload: Restore memory cache from catalog on DataCoord restart
+//
+// REFERENCE TRACKING:
+// - Track which snapshots reference each segment/index ID
+// - Enable safe GC by preventing deletion of segments/indexes in active snapshots
+// - Support queries like "which snapshots contain this segment?"
+//
+// CONCURRENCY:
+// - Thread-safe memory cache using ConcurrentMap
+// - Catalog operations are atomic via etcd transactions
+// - S3 operations use unique paths (collection_id/snapshot_id) to prevent conflicts
+
+// SnapshotDataInfo holds cached snapshot metadata with precomputed ID sets for fast lookup.
+//
+// This structure is used in the in-memory cache to enable efficient queries like:
+//   - Which snapshots contain segment X?
+//   - Which snapshots reference index Y?
+//
+// The ID sets are populated during snapshot creation and reload, avoiding the need
+// to scan segment/index lists on every query.
+type SnapshotDataInfo struct {
+	snapshotInfo *datapb.SnapshotInfo // Basic snapshot metadata (name, ID, collection, partitions, S3 location)
+	SegmentIDs   typeutil.UniqueSet   // Set of all segment IDs in this snapshot (for fast lookup)
+	IndexIDs     typeutil.UniqueSet   // Set of all index IDs in this snapshot (for fast lookup)
+}
+
+// snapshotMeta manages snapshot metadata both in memory and persistent storage.
+//
+// This is the central coordinator for snapshot operations, maintaining:
+//   - In-memory cache for fast lookup and reference tracking
+//   - Catalog persistence for durability across restarts
+//   - S3 reader/writer for complete snapshot data
+type snapshotMeta struct {
+	catalog metastore.DataCoordCatalog // Persistent storage in etcd/KV store for snapshot metadata
+
+	// In-memory cache: snapshot ID -> snapshot reference info
+	// Enables fast lookup and reference tracking (which snapshots contain a given segment/index)
+	// Thread-safe for concurrent access
+	snapshotID2DataInfo *typeutil.ConcurrentMap[UniqueID, *SnapshotDataInfo]
```

Comment:
need snapshotname and collection level index.

might need more index in the future. 

iterate every time is too slow